x/subkeys/genesis.go

```diff
@@ -0,0 +1,39 @@
+package subkeys
+
+import sdk "github.com/cosmos/cosmos-sdk/types"
+
+// GenesisState contains a set of fee allowances, persisted from the store
+type GenesisState []FeeAllowanceGrant
+
+// ValidateBasic ensures all grants in the genesis state are valid
+func (g GenesisState) ValidateBasic() error {
+	for _, f := range g {
+		err := f.ValidateBasic()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// InitGenesis will initialize the keeper from a *previously validated* GenesisState
+func InitGenesis(ctx sdk.Context, k Keeper, gen GenesisState) error {
+	for _, f := range gen {
+		err := k.DelegateFeeAllowance(ctx, f)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ExportGenesis will dump the contents of the keeper into a serializable GenesisState
+func ExportGenesis(ctx sdk.Context, k Keeper) (GenesisState, error) {
+	// TODO: all expiration heights will be thrown off if we dump state and start at a new
```

Comment:
I'd like to have a call with you on this one.
Given I am opposed to the (ab)use of this dump state and restart, and have an open upgrade PR to avoid the need to do so, I would like you all to give some direction.

My first thought would be to let the allowance subtract the block height at export, eg. prepare itself for export, given context. But let's discuss that - this is why there is a TODO

Reply:
Should we perhaps update the `ExportGenesis` interface to take in a block header?

Reply:
It is there in the context. I will add a second method to the Allowances so they know how to adjust over an export. And then dump them

Reply:
Resolved in 73eab922e